紀錄/24 20210406.md

```diff
@@ -0,0 +1,125 @@
+ 2. 目的: 線上一起寫題目, 由有想法的人帶領, 先解題, 再看該題有趣的解法
+ 3. 時間: 4/06 (二) 20:00 ~ 21:00
+ 4. 地點: google meet 線上 (前 10 分鐘預備鏈接)
+ 5. 解題項目:  [Binary search](https://leetcode.com/explore/learn/card/binary-search)
+ 6. 共筆: GitHub https://github.com/programmingbookclub/Leetcode-club
+ 
+ 
+ * Template II
+*   Article Binary Search Template II
+*     EASY	 278	 First Bad Version	 https://leetcode.com/problems/first-bad-version
+
+```Swift 
+// swift 
+// by nghuiqin
+func firstBadVersion(_ n: Int) -> Int {
+    if n == 0 { return -1 }
+    var low = 0
+    var high = n
+    while low < high {
+        let pivot = low + (high - low)/2
+        let isMidBad = isBadVersion(pivot)
+        if isMidBad == false {
+            low = pivot + 1
+        } else {
+            high = pivot
+        }
+    }
+    return high
+}
+*     MEDIUM	 162	 Find Peak Element	 https://leetcode.com/problems/find-peak-element
+```Swift
+func findPeakElement(_ nums: [Int]) -> Int {
+    var low = 0
+    var high = nums.count - 1
+    while low < high {
+        let pivot = (high - low)/2 + low
+        if nums[pivot] > nums[pivot+1] {
+            high = pivot
+        } else {
+            low = pivot + 1
+        }
+    }
+    return low
+}
+```swift
+// swift 
+// by danishuo
+class Solution {
+    func findPeakElement(_ nums: [Int]) -> Int {
+        var left = 0
+        var right = nums.count - 1
+        while(left < right) {
+            let mid = left + (right - left) / 2
+            if nums[mid+1] > nums[mid] {
+                left = mid + 1
+            } else {
+                right = mid
+            }
+        }
+        return left
+    }
+}
+```Swift
+// swift by SamTienz
+// SamTienz: 想說用recursive做看看 但發現好像只能抓到 Peak Element 不能找到當初的Index
+// edit by ytyubox: 使用 inner function 可以處理掉
+func findPeakElement(_ nums: [Int]) -> Int {
+        func helper(_ low: Int, _ high: Int) -> Int {
+            if low >= high {return low}
+            let pivot = (high - low) / 2 + low
+            if nums[pivot] > nums[pivot + 1] { 
+               return helper(low, pivot)
+            } else { 
+               return helper(pivot + 1, high)
+            }
+        }
+        return helper(0, nums.count - 1)
+    }
+```
+* MEDIUM	 153	 Find Minimum in Rotated Sorted Array	 https://leetcode.com/problems/find-minimum-in-rotated-sorted-array
+* Template III
+* Article Binary Search Template III
+MEDIUM    34    Find First and Last Position of Element in Sorted Array    https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
+
+
+```swift
+// swift by danishuotw 
+func searchRange(_ nums: [Int], _ target: Int) -> [Int] {        
+        var list = [-1, -1]
+        if let s = nums.firstIndex(of: target) { 
+            list[0] = s
+        }
+        if let e = nums.lastIndex(of: target) {
+            list[1] = e
+        }
+        return list
+    }
+```
+* MEDIUM	 658	 Find K Closest Elements	 https://leetcode.com/problems/find-k-closest-elements
```

Comment:
```suggestion
```typescript
// `Search for a Range` in TypeScript by iteration and resursion mixed
// Time Complexity: O(log n), Space Complexity: O(1)

function searchRange(nums: number[], target: number): number[] {
  let left = 0;
  let right = nums.length - 1;
  let middle: number;
  while (left <= right) {
    middle = getMiddleIndex(left, right);
    if (nums[middle] === target) {
      const leftTargerBoundary = searchLeftTargetBoundary(left, middle);
      const rightTargetBoundary = searchRightTargetBoundary(middle, right);

      const result = [leftTargerBoundary, rightTargetBoundary];
      return result;
    }
    else if (nums[middle] > target) right = middle - 1;
    else left = middle + 1;
  }
  return [-1, -1];
  
  // helper functions
  function getMiddleIndex(left: number, right: number) {
    if (left > right) return -1;
    if (left === right) return left;
    return left + Math.floor((right - left) / 2);
  }
  /** Search left boundary of target by Recursion */
  function searchLeftTargetBoundary(left: number, right: number) {
    if (nums[left] === target) return left;
    const middle = getMiddleIndex(left, right);
    
    if (nums[middle] === target && nums[middle - 1] !== target) return middle;
    else if (nums[middle] !== target) return searchLeftTargetBoundary(middle + 1, right);
    else return searchLeftTargetBoundary(left, middle - 1);
  }
  /** Search right boundary of target by Recursion */
  function searchRightTargetBoundary(left: number, right: number) {
    if (nums[right] === target) return right;
    const middle = getMiddleIndex(left, right);
    
    if (nums[middle] === target && nums[middle + 1] !== target) return middle;
    else if (nums[middle] !== target) return searchRightTargetBoundary(left, middle - 1);
    else return searchRightTargetBoundary(middle + 1, right);
  }
};
```

* MEDIUM	 658	 Find K Closest Elements	 https://leetcode.com/problems/find-k-closest-elements
```